### Libraries and initial data

In [ ]:
!pip install dnspython
!pip install pymongo[srv]

In [ ]:
import json #  to load data from json
import pandas as pd # for dataFrames
import seaborn as sns # for fancy plots


import pymongo # for db
import dns # DNS toolkit for python

In [ ]:
!curl ipecho.net/plain

35.231.54.111

Data is loaded from JSON, stored to the NoSQL database (MongoDB), retrieved, and stored to the DataFrame for the purpose of visualization.

In [ ]:
client = pymongo.MongoClient("mongodb+srv://julie_home:268902@cluster0-napnq.mongodb.net/maps?retryWrites=true&w=majority")
db = client["map"]
maps = db["coordinates"]

In [ ]:
data = [x for x in maps.find()] 
coords_big = data[0]['coordinates'][0][0]

## 2. Propose an algorithm that would break the map of Ukraine into identical squares (side ~ 1 km).

It would be easy and fast just to divide the range of latitude and longitude into a number of pieces. And it would even work well. If Earth were flat.

But it's not. So the distance between two meridians differs depending on the latitude.

So the algorithm is a bit more difficult.

Ukraine is a big country, so let's look at the data.

To calculate min/max coordinates of Ukraine:

In [ ]:
latitude_min = min([i[1] for i in coords_big])
latitude_max = max([i[1] for i in coords_big])

longitude_min = min([i[0] for i in coords_big])
longitude_max = max([i[0] for i in coords_big])

print('Latitude: {} - {}\nLongitude: {} - {}'.format(latitude_min, latitude_max, longitude_min, longitude_max))

Latitude: 44.1845979 - 52.3797464
Longitude: 22.1370589 - 40.2275801


Latitudes differ from 44.1845979 to 52.3797464 and longitude from 22.1370589 to 40.2275801.

In general, to calculate the actual length of the 1 degree on the map we need to use these formulas:
 - km_in_degree_longtitude = 111.321377778 * cos(latitude)
 - km_in_degree_latitude = 111.134861111

In [ ]:
km_in_degree_longitude = 111.321377778
km_in_degree_latitude = 111.134861111

In [ ]:
import math # for cos() and radians()

print('The difference of the length of 1 degree: ', end = '')
print(km_in_degree_longitude*(math.cos(math.radians(latitude_min))-math.cos(math.radians(latitude_max))))

The difference of the length of 1 degree: 11.8749628443405


And as we can see for different points on the map the actual km_in_degree_latitude is almost 12 km bigger.

In [ ]:
print('Latitude (constant): ',km_in_degree_latitude)
print('Longitude: from {} to {} km'.format(km_in_degree_longitude*math.cos(math.radians(latitude_min)), km_in_degree_longitude*math.cos(math.radians(latitude_max))))

Latitude (constant):  111.134861111
Longitude: from 79.8283364518405 to 67.9533736075 km


So we create a dictionary of the latitude of every row of squares. As the number of km in one degree of the latitude is constant that's pretty easy to do.

And the beginning of the dict looks like this:


In [ ]:
import numpy as np

In [ ]:
[i for i in np.arange(latitude_min, latitude_max, 1/km_in_degree_latitude)][:5]

[44.1845979,
 44.19359597666112,
 44.202594053322244,
 44.211592129983366,
 44.22059020664449]

If we look at the schematic map the highlighted values have been stored for every km:

![alt text](https://drive.google.com/uc?id=1KE-rcfKeHgbU4a2vMpteXzXXulIdM9rK)

With the range of longitude values, it's not so easy. As it was said the number of km in one degree depends on the latitude.
If you start dividing values from the longitude 22 for every latitude the map will look a bit like a **right** trapezoid (depicted below). Meanwhile, it's more logical to have it **isosceles** (or as close to it as possible).

 
 ![alt text](https://drive.google.com/uc?id=1ELNQ6aGD0pON8I0ig7jfaBUDXSy_LO5F)


So as depicted above for the Isosceles trapezoid, we choose longitude to make the map look symmetrical (again more or less) relative to that longitude. As line "b" in the picture.

I choose longitude 32 for this purpose. This is NOT even close to perfect, but more logical as for me.

So the visualization of what I'm about to do is:

![alt text](https://drive.google.com/uc?id=1Ems36gDIeV73WkCJ4DXGNYKQ1jbLwAg8)

As a result, we are going to get squares (that does not look like squares on this picture).



To save some memory and in general make the results less redundant I am about to keep ***only the coordinates of one corner of the square***. 

![alt text](https://drive.google.com/uc?id=1pidV4NaHMwtS5YZJtydzXanc3TEDaM3Q)

So for the square highlighted in the picture we are saving only coordinate of the left upper corner. All other coordinates are easily calculated with two simple formulas:
- for latitude of the lower corners = 52 + 1/km_in_degree_latitude; 
- for longitude of the right corners = 32 + 1/(km_in_degree_latitude * cos(latitude))

In [ ]:
def closest_smaller(coord, km_1, map_squares):
  '''
  find the closest smaller value in the dictionary

  Parameters:
  coord - the value that we want to "round"
  km_1 = 1/km_in_degree
  '''
  for i in map_squares:
    if i>coord:
      return i-km_1

In [ ]:
'''
create an array from the initial coords_big where each latitude is rounded 
to the closest smaller value in the dictionary 
and create a dictionary of all the coordinates
'''
coords_latitude_cut = []
for i in coords_big:
  coords_latitude_cut.append([i[0], closest_smaller(i[1], 1/km_in_degree_latitude, map_squares)])

map_squares_cut = {i:[] for i in np.arange(latitude_min, latitude_max, 1/km_in_degree_latitude)}

for i in coords_latitude_cut:
  if not i[1] is None:
    map_squares_cut[i[1]].append(i[0])

In [ ]:
map_squares_ranges = {i:[] for i in np.arange(latitude_min, latitude_max, 1/km_in_degree_latitude)}

temp = [coords_latitude_cut[0]]
for i in range(len(coords_latitude_cut)):
  if temp[-1][1]==coords_latitude_cut[i][1] and not temp[-1][0]==coords_latitude_cut[i][0]:
    temp.append(coords_latitude_cut[i])
  else:
    if not coords_latitude_cut[i][1] is None:
      map_squares_ranges[coords_latitude_cut[i-1][1]].append((np.average([i[0] for i in temp])))
      temp = [coords_latitude_cut[i]]
if not coords_latitude_cut[i][1] is None:
  map_squares_ranges[coords_latitude_cut[i-1][1]].append((np.average([i[0] for i in temp])))  


for i in map_squares_ranges.keys():
  map_squares_ranges[i] = sorted(map_squares_ranges[i])

In [ ]:
for i in map_squares_ranges.keys():
  if len(map_squares_ranges[i])%2 == 1 and not i == latitude_min:
    map_squares_ranges[i]=map_squares_ranges[i-1/km_in_degree_latitude]
  map_squares_ranges[i] = np.array(map_squares_ranges[i]).reshape((-1,2))

In [ ]:
def get_longitude_range(latitude, longitude_range, middle = 32):
  '''
  Function for getting a range of the longitudes for the specific latitude.
  In the list returned ALL values belong to the polygon (map).
  The distance between the two values is approximately 1 km.

  Parameters:
  latitude - the latitude for which the range will be calculated
  middle (optional) - the axis of symmetry (default = 32)
  '''

  # range of values after "middle" value
  max_longitude = longitude_range[-1][-1]
  after_middle = np.arange(middle, max_longitude, 1/(math.cos(math.radians(latitude)) * km_in_degree_longitude))
  
  # range of values before "middle" value in reverse order
  min_longitude = longitude_range[0][0]
  before_middle = np.arange(middle, min_longitude, -1/(math.cos(math.radians(latitude)) * km_in_degree_longitude))

  # all values that lies between min and max longitude
  longitude_range_full = np.concatenate((before_middle[::-1], after_middle))


  # adding to the final list only values that belong to the polygon
  longitude_final_range = []
  for i in longitude_range_full:
    if any([i > j[0] and i < j[1] for j in longitude_range]):
      longitude_final_range.append(i)
  
  if longitude_final_range == []:
    longitude_final_range = longitude_range
  return longitude_final_range

To illustrate:

For the latitude of 49.997 first and last 5 values:

In [ ]:
range_example = get_longitude_range(49.997355423084784, map_squares_ranges[49.997355423084784])

In [ ]:
map_squares_ranges[49.997355423084784]

array([[23.18247863, 37.9501119 ],
       [38.203395  , 38.3677996 ],
       [38.44337748, 38.6789949 ],
       [38.69006657, 38.69583133]])

In [ ]:
print(range_example[:5])
print(range_example[-10:])

[23.19619140955816, 23.210165708908068, 23.224140008257976, 23.238114307607884, 23.25208860695779]
[38.55394639510837, 38.56792069445828, 38.58189499380819, 38.5958692931581, 38.609843592508014, 38.623817891857925, 38.637792191207836, 38.65176649055775, 38.66574078990766, 38.69368938860748]


On the map (copied from earlier calculations) we can see that indeed on latitude 49.997 the map starts approximately 23 degrees. And by the end of the map on this latitude, there are 2 tiny overhung parts (the tiniest is in the circle) of the map that is displayed in the last values.  

..., 38.55394639510837, 38.56792069445828)----(38.58189499380819, ... 38.66574078990766)-----(38.69368938860748)


![alt text](https://drive.google.com/uc?id=1MgThJ1_b1s6Qy09d2gZ_VC_av5Xvk5Qs)

So the algorithm works.

In [ ]:
map_squares = {i:[] for i in np.arange(latitude_min, latitude_max, 1/km_in_degree_latitude)}

In [ ]:
%%time
for i in map_squares:
  if len(map_squares_ranges[i]) == 0:
    pass
  else:
    map_squares[i] = get_longitude_range(i, map_squares_ranges[i])

CPU times: user 1.37 s, sys: 5.77 ms, total: 1.38 s
Wall time: 1.38 s


In [ ]:
for i in map_squares_ranges:
  if len(map_squares[i]) == 0:
    map_squares[i] = map_squares[i-1/km_in_degree_latitude]

In [ ]:
len(list(map_squares.keys()))

911

The map was divided by the latitude to the 911 pieces. Official statistic claims that the size of Ukraine from the North to the South is 893 - 896 km. Which tells us that the error of the latitude division algorithm is from 1.9758% to 1.6465%.

In [ ]:
max([len(map_squares[i]) for i in map_squares])

1304

Actual width of Ukraine (from East to West) is approximately 1316 km.
Maximum width in out algorithm is 1304.

The error is 0.912%

### To get a coordinate of a [random] square

In [ ]:
import random

random_latitude_lu = random.choice(list(map_squares.keys()))
random_longitude_lu = random.choice(map_squares[random_latitude_lu])

print(random_longitude_lu,random_latitude_lu)

29.49691560217542 48.674638153899856


In [ ]:
random_longitude_lu=29.49691560217542
random_latitude_lu=48.674638153899856

The point somewhere not far from Uman'.

For working with the squares I'm goint to use following denotation:

In [ ]:
"""
    LU -------------- RU
(left upper)     (right upper)
    |                  |
    |                  |
    |                  |
    |                  |
    |                  |
    LL --------------- RL
(left lower)     (right lower)
"""

In [ ]:
def get_square_coords(random_longitude_lu, random_latitude_lu):
  random_coords = {}
  random_coords['lu'] = (random_longitude_lu,random_latitude_lu)
  random_coords['ru'] = (random_longitude_lu + 1/(km_in_degree_latitude * math.cos(math.radians(random_latitude_lu))),random_latitude_lu)
  random_coords['ll'] = (random_longitude_lu,random_latitude_lu + 1/km_in_degree_latitude)
  random_coords['rl'] = (random_longitude_lu + 1/(km_in_degree_latitude * math.cos(math.radians(random_latitude_lu + 1/km_in_degree_latitude))),random_latitude_lu + 1/km_in_degree_latitude)
  return(random_coords)

In [ ]:
get_square_coords(random_longitude_lu, random_latitude_lu)

{'ll': (29.49691560217542, 48.68363623056098),
 'lu': (29.49691560217542, 48.674638153899856),
 'rl': (29.51054458706836, 48.68363623056098),
 'ru': (29.510542152732555, 48.674638153899856)}

And final square with these coordinates looks like this:

![square_sample](https://drive.google.com/uc?id=1tN7WNVxONWW5O5dUb6knA3qw4dpU0gXR)

With area of the square equals to 1 km^2 and perimeter equals to 4 km.

## 3. Save the coordinates of the formed squares.

In [ ]:
squares = db["squares"]

In [ ]:
code = 0
squares_list = []
for latitude in list(map_squares):
  for longitude in map_squares[latitude]:
    squares_list.append({'id':code, 'latitude': latitude, 'longitude':longitude})
    code += 1

In [ ]:
squares.insert_many(squares_list)